### ISSUE TITLE
PostCSS can't parse CSS variable values

### ISSUE THREAD

**Issue by tomhodgins** [postcss/postcss#1255]
It appears PostCSS has difficulty parsing CSS variables according to CSS's syntax, examples include:

```css
examples {
  --bracket-block: {1, 2, 3};
  --JSON: [1, "2", {"three": 3}, [4]];
  --javascript: function(rule) { console.log(rule) };
}
```

You can find more known-valid examples of CSS variables in my [css-tests](https://github.com/tomhodgins/css-tests/blob/master/value/css-variable.css) repository, or in the [Web Platform Tests repository](https://github.com/web-platform-tests/wpt/blob/master/css/css-variables/test_variable_legal_values.html#L64&L86), and you can read about the syntax for CSS variable values in the [CSS Variables spec](https://www.w3.org/TR/css-variables-1/#syntax)

> The allowed syntax for custom properties is extremely permissive. The `<declaration-value>` production matches any sequence of one or more tokens, so long as the sequence does not contain `<bad-string-token>`, `<bad-url-token>`, unmatched `<)-token>`, `<]-token>`, or `<}-token>`, or top-level `<semicolon-token>` tokens or `<delim-token>` tokens with a value of "`!`".

**Comment by ai** [comment]
Your examples contains `<}-token>`. Are you sure, that it is valid?

**Comment by evilebottnawi** [comment]
@ai yes, it is valid for css variables

**Comment by evilebottnawi** [comment]
It doesn't used in `css` in real life, but css spec says that in the future they can be used

**Comment by ai** [comment]
> so long as the sequence does not contain … <}-token>

**Comment by evilebottnawi** [comment]
hm, chrome and firefox parse and output as value of variable

**Comment by tomhodgins** [comment]
@ai does it contain _unmatched_ `<}-token>`? The examples I've given are valid according to:

- the spec
- all browser's implementations of CSS according to the specs
- the web platform tests that assert these are valid

And @evilebottnawi with the Houdini APIs coming, CSS variable can be used to pass values from CSS to JS, the syntax for CSS variables being so permissive was intentional. Here's an example of that: https://houdini.glitch.me/paint

**Comment by ai** [comment]
Got it. OK I will fix it in next month.

**Comment by ai** [comment]
I created a task on Evil Martians’s open source task list for junior developers https://cultofmartians.com/tasks/postcss-vars-parse.html

**Comment by stas-dolgachov** [comment]
@ai @tomhodgins hey! I have started to work on that. I am new to PostCSS source code, so it could take a while. I will give an update in a week or so.

**Comment by ai** [comment]
@stas-dolgachov great! Don’t worry to ask questions. You can write to me to `@sitnik` at Telegram.

**Comment by tomhodgins** [comment]
I've made [a little tool here](https://tomhodgins.com/demo/cssom/) that if you enter CSS into the editor on the left or top of the page, the browser will try to parse the CSS and the result is visible on the right or bottom of the page. This is to be able to instantly test how each browser 'sees' some CSS code. For troubleshooting CSS variables I haven't noticed any cross-browser differences in parsing (I believe they all do it according to the standard) so hopefully this will help you test valid values faster as you troubleshoot!

**Comment by ai** [comment]
Fixed in https://github.com/postcss/postcss/pull/1259

**Comment by ai** [comment]
The fix was released in 7.0.15.

**Comment by ai** [comment]
7.0.15 was reverted in 7.0.16.

Custom Properties fix will be released only in 8.0 until `postcss-apply` will be updated.
